src/daemon.rs

```diff
@@ -138,3 +145,238 @@ impl fmt::Display for NotifyState {
+/// https://www.freedesktop.org/software/systemd/man/systemd.socket.html
+#[derive(Debug, Clone)]
+pub enum SocketType {
+    /// A FIFO named pipe (see man 7 fifo)
+    Fifo(RawFd),
+    /// A special file, such as character device nodes or special files in 
+    /// /proc and /sys
+    Special(RawFd),
+    /// A PF_INET socket, such as UDP/TCP sockets
+    Inet(RawFd),
+    /// A PF_UNIX socket (see man 7 unix)
+    Unix(RawFd),
+    /// A POSIX message queue (see man 7 mq_overview)
+    Mq(RawFd),
+}
+
+impl SocketType {
+    /// Returns true if SocketType is a PF_UNIX socket
+    pub fn is_unix(&self) -> bool {
+        match self {
+            SocketType::Unix(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Returns true if SocketType is a PF_INET socket
+    pub fn is_inet(&self) -> bool {
+        match self {
+            SocketType::Inet(_) => true,
+            _ => false,
+        }
+    }
+}
+
+pub fn sd_booted() -> bool {
+    unimplemented!();
+}
+
+/// Check for file descriptors passed by systemd
+///
+/// Invoked by socket activated daemons to check for file descriptors needed by the service.
+/// If unset_env is true, the environment variables used by systemd will be cleared.
+pub fn sd_listen_fds(unset_env: bool) -> Result<Vec<SocketType>> {
+    let pid = env::var("LISTEN_PID").expect("LISTEN_PID");
+    let pid = pid.parse::<u32>().unwrap();
+    if process::id() != pid {
+        return Err("Pid mismatch".into());
+    }
+
+    let fds = env::var("LISTEN_FDS").expect("LISTEN_FDS");
+    let fds = fds.parse::<i32>().unwrap();
+
+    if unset_env {
+        env::remove_var("LISTEN_PID");
+        env::remove_var("LISTEN_FDS");
+        env::remove_var("LISTEN_FDNAMES");
+    }
+
+    let vec = socks_from_fds(fds);
+    Ok(vec)
+}
+
+/// Check for file descriptors passed by systemd
+///
+/// Like sd_listen_fds, but will also return a Vec of names associated with each file
+/// descriptor.
+pub fn sd_listen_fds_with_names(unset_env: bool) -> Result<Vec<(SocketType, String)>> {
+    let pid = env::var("LISTEN_PID").expect("LISTEN_PID");
+    let pid = pid.parse::<u32>().unwrap();
+    if process::id() != pid {
+        return Err("Pid mismatch".into());
+    }
+
+    let fds = env::var("LISTEN_FDS").expect("LISTEN_FDS");
+    let fds = fds.parse::<i32>().unwrap();
+
+    let names = env::var("LISTEN_FDNAMES").expect("LISTEN_FDNAMES");
+
+    let names: Vec<String> = names.split(":").map(String::from).collect();
+    let vec = socks_from_fds(fds);
+
+    let out = vec.into_iter().zip(names.into_iter()).collect();
+
+    if unset_env {
+        env::remove_var("LISTEN_PID");
+        env::remove_var("LISTEN_FDS");
+        env::remove_var("LISTEN_FDNAMES");
+    }
+
+    Ok(out)
+}
+
+fn socks_from_fds(fds: RawFd) -> Vec<SocketType> {
+    let mut vec = Vec::new();
+    for fd in SD_LISTEN_FDS_START..SD_LISTEN_FDS_START+fds {
+        let sock = SocketType::try_from(fd).expect("Socket type conversion");
```

Comment:
I think we should either return a `Result` here, or log and skip when `try_from` fails.